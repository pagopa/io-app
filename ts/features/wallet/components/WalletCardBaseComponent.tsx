import { WithTestID } from "@pagopa/io-app-design-system";
import React from "react";
import { StyleSheet } from "react-native";
import Animated, {
  FadeIn,
  FadeOut,
  LinearTransition
} from "react-native-reanimated";

// Wallet card base component props, which declares common props that wallet cards must have
export type WalletCardComponentBaseProps<P> = WithTestID<{
  isStacked?: boolean;
  cardProps: P;
}>;

export const withWalletCardBaseComponent =
  <
    CardProps extends object,
    ContentProps extends WalletCardComponentBaseProps<CardProps>
  >(
    CardContent: React.ComponentType<CardProps>
  ) =>
  ({ cardProps, isStacked = true, testID }: ContentProps) =>
    (
      <Animated.View
        testID={testID}
        style={[styles.container, isStacked && styles.containerStacked]}
        layout={LinearTransition.duration(200)}
        entering={FadeIn.duration(200)}
        exiting={FadeOut.duration(200)}
      >
        <CardContent {...cardProps} />
      </Animated.View>
    );

export type WalletCardBaseComponent<
  CardProps extends object = object,
  ContentProps extends WalletCardComponentBaseProps<CardProps> = WalletCardComponentBaseProps<CardProps>
> = ReturnType<typeof withWalletCardBaseComponent<CardProps, ContentProps>>;

const styles = StyleSheet.create({
  container: {
    aspectRatio: 16 / 10
  },
  containerStacked: {
    aspectRatio: 16 / 3
  }
});
