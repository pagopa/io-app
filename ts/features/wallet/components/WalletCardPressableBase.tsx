import { PressableBaseProps, WithTestID } from "@pagopa/io-app-design-system";
import * as React from "react";
import { Pressable } from "react-native";
import Animated from "react-native-reanimated";
import { useSpringPressScaleAnimation } from "../../../components/ui/utils/hooks/useSpringPressScaleAnimation";

export type WalletCardPressableBaseProps = WithTestID<PressableBaseProps>;

export const WalletCardPressableBase = ({
  onPress,
  testID,
  accessibilityLabel,
  children
}: React.PropsWithChildren<WalletCardPressableBaseProps>) => {
  const { onPressIn, onPressOut, animatedScaleStyle } =
    useSpringPressScaleAnimation();

  if (onPress === undefined) {
    return <>{children}</>;
  }

  return (
    <Pressable
      onPress={onPress}
      testID={testID}
      accessible={true}
      accessibilityLabel={accessibilityLabel}
      accessibilityRole="button"
      onPressIn={onPressIn}
      onPressOut={onPressOut}
      onTouchEnd={onPressOut}
    >
      <Animated.View style={animatedScaleStyle}>{children}</Animated.View>
    </Pressable>
  );
};
