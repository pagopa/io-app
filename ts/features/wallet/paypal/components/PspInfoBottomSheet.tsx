import { View } from "native-base";
import * as React from "react";
import { StyleSheet } from "react-native";
import { NonNegativeNumber } from "@pagopa/ts-commons/lib/numbers";
import I18n from "../../../../i18n";
import { IOColors } from "../../../../components/core/variables/IOColors";
import ButtonDefaultOpacity from "../../../../components/ButtonDefaultOpacity";
import { openWebUrl } from "../../../../utils/url";
import { Link } from "../../../../components/core/typography/Link";
import { useIOBottomSheet } from "../../../../utils/bottomSheet";
import { Body } from "../../../../components/core/typography/Body";

const styles = StyleSheet.create({
  link: {
    backgroundColor: IOColors.white,
    borderColor: IOColors.white,
    paddingRight: 0,
    paddingLeft: 0
  }
});

const findOutMore = "https://io.italia.it/cashback/faq/#n3_11";

/**
 * Explains why there are other cards
 * @constructor
 */
const WhyOtherCards = () => (
  <View>
    <View spacer={true} />
    <View style={{ flex: 1 }}>
      <Body>
        {I18n.t(
          "bonus.bpd.details.paymentMethods.activateOnOthersChannel.whyOtherCards.body"
        )}
      </Body>
      <ButtonDefaultOpacity
        onPress={() => openWebUrl(findOutMore)}
        onPressWithGestureHandler={true}
        style={styles.link}
      >
        <Link>
          {I18n.t(
            "bonus.bpd.details.paymentMethods.activateOnOthersChannel.whyOtherCards.cta"
          )}
        </Link>
      </ButtonDefaultOpacity>
    </View>
  </View>
);

type Props = {
  pspName: string;
  pspFee: NonNegativeNumber;
};
export const usePspInfoBottomSheet = (props: Props) =>
  useIOBottomSheet(
    <WhyOtherCards />,
    I18n.t("wallet.onboarding.paypal.selectPsp.infoBottomSheet.title", {
      pspName: props.pspName
    }),
    300
  );
