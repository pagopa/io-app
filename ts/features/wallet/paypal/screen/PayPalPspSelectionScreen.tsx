import React, { useState } from "react";
import {
  Dimensions,
  Image,
  SafeAreaView,
  ScrollView,
  StyleSheet
} from "react-native";
import { Dispatch } from "redux";
import { connect } from "react-redux";
import { View } from "native-base";
import { NonNegativeNumber } from "@pagopa/ts-commons/lib/numbers";
import BaseScreenComponent from "../../../../components/screens/BaseScreenComponent";
import { emptyContextualHelp } from "../../../../utils/emptyContextualHelp";
import I18n from "../../../../i18n";
import { IOStyles } from "../../../../components/core/variables/IOStyles";
import { GlobalState } from "../../../../store/reducers/types";
import { H1 } from "../../../../components/core/typography/H1";
import { Body } from "../../../../components/core/typography/Body";
import { Link } from "../../../../components/core/typography/Link";
import FooterWithButtons from "../../../../components/ui/FooterWithButtons";
import { openWebUrl } from "../../../../utils/url";
import {
  RadioButtonList,
  RadioItem
} from "../../../../components/core/selection/RadioButtonList";
import { privacyUrl } from "../../../../config";
import { isReady, remoteReady } from "../../../bonus/bpd/model/RemoteValue";
import { useImageResize } from "../../onboarding/bancomat/screens/hooks/useImageResize";
import { H4 } from "../../../../components/core/typography/H4";
import IconFont from "../../../../components/ui/IconFont";
import { IOColors } from "../../../../components/core/variables/IOColors";
import TouchableDefaultOpacity from "../../../../components/TouchableDefaultOpacity";

type PayPalPsp = {
  id: string;
  logoUrl: string;
  name: string;
  fee: NonNegativeNumber;
  privacyUrl: string;
};
const IMAGE_WIDTH = Dimensions.get("window").width;
const IMAGE_HEIGHT = 32;
const styles = StyleSheet.create({
  pspLogo: {
    height: 32,
    flex: 1,
    flexDirection: "row",
    justifyContent: "flex-start"
  }
});

const pspList: ReadonlyArray<PayPalPsp> = [
  {
    id: "1",
    logoUrl: "https://paytipper.com/wp-content/uploads/2021/02/logo.png",
    name: "PayTipper",
    fee: 100 as NonNegativeNumber,
    privacyUrl
  },
  {
    id: "2",
    logoUrl:
      "https://www.pikpng.com/pngl/m/600-6000675_business-wire-logo-png-intesa-sanpaolo-clipart.png",
    name: "Intesa San Paolo",
    fee: 200 as NonNegativeNumber,
    privacyUrl
  }
];

const RadioItemBody = ({ psp }: { psp: PayPalPsp }) => {
  const imgDimensions = useImageResize(IMAGE_WIDTH, IMAGE_HEIGHT, psp.logoUrl);

  return (
    <View
      style={{
        alignItems: "flex-start",
        justifyContent: "space-between",
        flexDirection: "row"
      }}
    >
      {imgDimensions.fold<React.ReactNode>(<H4>psp.name</H4>, imgDim => (
        <Image
          source={{ uri: psp.logoUrl }}
          style={[styles.pspLogo, { width: imgDim[0], height: imgDim[1] }]}
          resizeMode={"contain"}
        />
      ))}
      <View
        style={{
          flex: 1,
          flexDirection: "column",
          alignItems: "flex-end"
        }}
      >
        <TouchableDefaultOpacity
          onPress={() => console.log("pressed")}
          style={{
            flex: 1,
            flexDirection: "column",
            alignItems: "flex-end"
          }}
        >
          <IconFont name={"io-info"} size={24} color={IOColors.blue} />
        </TouchableDefaultOpacity>
      </View>
    </View>
  );
};

const getPspListRadioItems = (
  pspList: ReadonlyArray<PayPalPsp>
): ReadonlyArray<RadioItem<PayPalPsp["id"]>> =>
  pspList.map(psp => ({
    id: psp.id,
    body: {
      kind: "node",
      element: <RadioItemBody psp={psp} />
    }
  }));

type Props = ReturnType<typeof mapDispatchToProps> &
  ReturnType<typeof mapStateToProps>;

const getLocales = () => ({
  title: I18n.t("bonus.bpd.iban.insertion.title"),
  body: I18n.t("bonus.bpd.iban.insertion.body1"),
  bodyLinkTitle: "a link title"
});

/**
 * This screen is the start point to onboard PayPal as payment method
 * It shows the PP logo and some texts
 * At the bottom 2 CTA to cancel or continue
 */
const PayPalPpsSelectionScreen = (props: Props): React.ReactElement | null => {
  const { title, body, bodyLinkTitle } = getLocales();
  const [selectedPsp, setSelectedPsp] = useState<PayPalPsp["id"]>("");
  const cancelButtonProps = {
    testID: "cancelButtonId",
    primary: false,
    bordered: true,
    // TODO replace with the effective handler
    onPress: undefined,
    title: I18n.t("global.buttons.cancel")
  };
  const continueButtonProps = {
    testID: "continueButtonId",
    bordered: false,
    // TODO replace with the effective handler
    onPress: undefined,
    title: I18n.t("global.buttons.continue")
  };
  return (
    <BaseScreenComponent
      goBack={true}
      contextualHelp={emptyContextualHelp}
      headerTitle={I18n.t("wallet.onboarding.paypal.headerTitle")}
    >
      <SafeAreaView style={IOStyles.flex} testID={"PayPalPpsSelectionScreen"}>
        <View style={[IOStyles.horizontalContentPadding, IOStyles.flex]}>
          <View spacer={true} large={true} />
          <H1>{title}</H1>
          <View spacer={true} large={true} />
          <Body>{body}</Body>
          <Link onPress={() => openWebUrl("https://www.google.com")}>
            {bodyLinkTitle}
          </Link>
          <View spacer={true} large={true} />

          {isReady(props.pspList) && (
            <ScrollView>
              <View style={{ flexDirection: "row" }}>
                <H4 color={"bluegreyDark"} weight={"Regular"}>
                  {"ciao"}
                </H4>
                <H4
                  color={"bluegreyDark"}
                  weight={"Regular"}
                  style={{
                    flex: 1,
                    textAlign: "right"
                  }}
                >
                  {"zio"}
                </H4>
              </View>

              <RadioButtonList<PayPalPsp["id"]>
                key="paypal_psp_selection"
                items={getPspListRadioItems(props.pspList.value)}
                selectedItem={selectedPsp}
                onPress={setSelectedPsp}
              />
            </ScrollView>
          )}
        </View>
        <FooterWithButtons
          type={"TwoButtonsInlineThird"}
          leftButton={cancelButtonProps}
          rightButton={continueButtonProps}
        />
      </SafeAreaView>
    </BaseScreenComponent>
  );
};

const mapDispatchToProps = (_: Dispatch) => ({});
const mapStateToProps = (_: GlobalState) => ({
  pspList: remoteReady(pspList)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(PayPalPpsSelectionScreen);
