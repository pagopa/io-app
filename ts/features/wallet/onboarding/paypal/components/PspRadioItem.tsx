// component that represents the item in the radio list
import * as O from "fp-ts/lib/Option";
import { pipe } from "fp-ts/lib/function";
import React from "react";
import { Dimensions, Image, StyleSheet, View } from "react-native";
import {
  ButtonSolid,
  Icon,
  VSpacer,
  ContentWrapper
} from "@pagopa/io-app-design-system";
import TouchableDefaultOpacity from "../../../../../components/TouchableDefaultOpacity";
import { H4 } from "../../../../../components/core/typography/H4";
import I18n from "../../../../../i18n";
import { TestID } from "../../../../../types/WithTestID";
import { useIOBottomSheetAutoresizableModal } from "../../../../../utils/hooks/bottomSheet";
import { useImageResize } from "../../bancomat/hooks/useImageResize";
import { IOPayPalPsp } from "../types";
import { PspInfoBottomSheetContent } from "./PspInfoBottomSheet";

export const PSP_LOGO_MAX_WIDTH = Dimensions.get("window").width;
export const PSP_LOGO_MAX_HEIGHT = 32;
type RadioItemProps = {
  psp: IOPayPalPsp;
} & TestID;

const styles = StyleSheet.create({
  pspLogo: {
    height: 32,
    flex: 1,
    flexDirection: "row",
    justifyContent: "flex-start"
  },
  radioItemBody: {
    alignItems: "flex-start",
    justifyContent: "space-between",
    flexDirection: "row"
  },
  radioItemRight: {
    flex: 1,
    flexDirection: "column",
    alignItems: "flex-end"
  }
});

/**
 * item used in a radio list
 * it represents a PayPal psp with logo and info icon
 * @param props
 * @constructor
 */
export const PspRadioItem = (
  props: RadioItemProps
): React.ReactElement | null => {
  const { psp } = props;
  const imgDimensions = useImageResize(
    PSP_LOGO_MAX_WIDTH,
    PSP_LOGO_MAX_HEIGHT,
    psp.logoUrl
  );

  const { present, bottomSheet, dismiss } = useIOBottomSheetAutoresizableModal(
    {
      title: I18n.t(
        "wallet.onboarding.paypal.selectPsp.infoBottomSheet.title",
        {
          pspName: psp.name
        }
      ),
      component: (
        <PspInfoBottomSheetContent
          pspFee={psp.fee}
          pspName={psp.name}
          pspPrivacyUrl={psp.privacyUrl}
        />
      ),
      footer: (
        <ContentWrapper>
          <ButtonSolid
            testID="continueButtonId"
            onPress={() => dismiss()}
            label={I18n.t(
              "wallet.onboarding.paypal.selectPsp.infoBottomSheet.ctaTitle"
            )}
            accessibilityLabel={I18n.t(
              "wallet.onboarding.paypal.selectPsp.infoBottomSheet.ctaTitle"
            )}
            fullWidth={true}
          />
          <VSpacer size={16} />
        </ContentWrapper>
      )
    },
    130
  );

  return (
    <View style={styles.radioItemBody} testID={props.testID}>
      {/* show the psp name while its image is loading */}
      {pipe(
        imgDimensions,
        O.fold(
          () => (
            <H4
              weight={"SemiBold"}
              color={"bluegreyDark"}
              testID={"pspNameTestID"}
            >
              {psp.name}
            </H4>
          ),
          imgDim => (
            <Image
              testID={"pspNameLogoID"}
              source={{ uri: psp.logoUrl }}
              style={[styles.pspLogo, { width: imgDim[0], height: imgDim[1] }]}
              resizeMode={"contain"}
            />
          )
        )
      )}
      <View style={styles.radioItemRight}>
        <TouchableDefaultOpacity testID={"infoIconTestID"} onPress={present}>
          <Icon name="info" size={24} color="blue" />
        </TouchableDefaultOpacity>
      </View>
      {bottomSheet}
    </View>
  );
};
