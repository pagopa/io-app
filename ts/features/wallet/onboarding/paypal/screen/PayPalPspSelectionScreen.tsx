import { useNavigation } from "@react-navigation/native";
import React, { useEffect, useState } from "react";
import { SafeAreaView, ScrollView, StyleSheet, View } from "react-native";
import { connect, useDispatch } from "react-redux";
import { Dispatch } from "redux";
import { VSpacer } from "@pagopa/io-app-design-system";
import {
  RadioButtonList,
  RadioItem
} from "../../../../../components/core/selection/RadioButtonList";
import { Body } from "../../../../../components/core/typography/Body";
import { H1 } from "../../../../../components/core/typography/H1";
import { H4 } from "../../../../../components/core/typography/H4";
import { Link } from "../../../../../components/core/typography/Link";
import { IOStyles } from "../../../../../components/core/variables/IOStyles";
import BaseScreenComponent from "../../../../../components/screens/BaseScreenComponent";
import FooterWithButtons from "../../../../../components/ui/FooterWithButtons";
import I18n from "../../../../../i18n";
import { GlobalState } from "../../../../../store/reducers/types";
import { emptyContextualHelp } from "../../../../../utils/emptyContextualHelp";
import { useIOBottomSheetAutoresizableModal } from "../../../../../utils/hooks/bottomSheet";
import { LoadingErrorComponent } from "../../../../../components/LoadingErrorComponent";
import {
  getValueOrElse,
  isError,
  isReady
} from "../../../../../common/model/RemoteValue";
import { PspRadioItem } from "../components/PspRadioItem";
import {
  searchPaypalPsp as searchPaypalPspAction,
  walletAddPaypalBack,
  walletAddPaypalCancel,
  walletAddPaypalPspSelected
} from "../store/actions";
import { payPalPspSelector } from "../store/reducers/searchPsp";
import { IOPayPalPsp } from "../types";
import PAYPAL_ROUTES from "../navigation/routes";

type Props = ReturnType<typeof mapDispatchToProps> &
  ReturnType<typeof mapStateToProps>;

const styles = StyleSheet.create({
  radioListHeaderRightColumn: {
    flex: 1,
    textAlign: "right"
  }
});

// an header over the psp list with 2 columns
const RadioListHeader = (props: {
  leftColumnTitle: string;
  rightColumnTitle: string;
}) => {
  const color = "bluegrey";
  const weight = "Regular";
  return (
    <View style={{ flexDirection: "row" }}>
      <H4 color={color} weight={weight}>
        {props.leftColumnTitle}
      </H4>
      <H4
        color={color}
        weight={weight}
        style={styles.radioListHeaderRightColumn}
      >
        {props.rightColumnTitle}
      </H4>
    </View>
  );
};

const getPspListRadioItems = (
  pspList: ReadonlyArray<IOPayPalPsp>
): ReadonlyArray<RadioItem<IOPayPalPsp["id"]>> =>
  pspList.map(psp => ({
    id: psp.id,
    body: <PspRadioItem psp={psp} />
  }));

const getLocales = () => ({
  title: I18n.t("wallet.onboarding.paypal.selectPsp.title"),
  body: I18n.t("wallet.onboarding.paypal.selectPsp.body"),
  link: I18n.t("wallet.onboarding.paypal.selectPsp.link"),
  leftColumnTitle: I18n.t("wallet.onboarding.paypal.selectPsp.leftColumnTitle"),
  rightColumnTitle: I18n.t(
    "wallet.onboarding.paypal.selectPsp.rightColumnTitle"
  ),
  whatIsPspBody: I18n.t(
    "wallet.onboarding.paypal.selectPsp.whatIsPspBottomSheet.body"
  ),
  whatIsPspTitle: I18n.t(
    "wallet.onboarding.paypal.selectPsp.whatIsPspBottomSheet.title"
  )
});

/**
 * This screen is where the user picks a PSP that will be used to handle PayPal transactions within PayPal
 * Only 1 psp can be selected
 */
const PayPalPspSelectionScreen = (props: Props): React.ReactElement | null => {
  const locales = getLocales();
  const { present: presentWhatIsPspBottomSheet, bottomSheet } =
    useIOBottomSheetAutoresizableModal({
      title: locales.whatIsPspTitle,
      component: (
        <View>
          <Body>{locales.whatIsPspBody}</Body>
          <VSpacer size={16} />
        </View>
      )
    });
  const pspList = getValueOrElse(props.pspList, []);
  const [selectedPsp, setSelectedPsp] = useState<IOPayPalPsp | undefined>();
  const dispatch = useDispatch();
  const navigation = useNavigation();
  const searchPaypalPsp = () => {
    dispatch(searchPaypalPspAction.request());
  };
  useEffect(searchPaypalPsp, [dispatch]);
  useEffect(() => {
    // auto select if the psp list has 1 element
    setSelectedPsp(pspList.length === 1 ? pspList[0] : undefined);
  }, [pspList]);

  const buttonsProps = {
    cancelButtonProps: {
      testID: "cancelButtonId",
      primary: false,
      bordered: true,
      onPress: props.cancel,
      title: I18n.t("global.buttons.cancel")
    },
    continueButtonProps: {
      testID: "continueButtonId",
      bordered: false,
      onPress: () => {
        if (selectedPsp) {
          props.setPspSelected(selectedPsp);
          navigation.navigate(PAYPAL_ROUTES.ONBOARDING.MAIN, {
            screen: PAYPAL_ROUTES.ONBOARDING.CHECKOUT
          });
        }
      },
      title: I18n.t("global.buttons.continue")
    }
  };

  return (
    <BaseScreenComponent
      goBack={props.goBack}
      contextualHelp={emptyContextualHelp}
      headerTitle={I18n.t("wallet.onboarding.paypal.headerTitle")}
    >
      {isReady(props.pspList) ? (
        <SafeAreaView style={IOStyles.flex} testID={"PayPalPpsSelectionScreen"}>
          <View style={[IOStyles.horizontalContentPadding, IOStyles.flex]}>
            <VSpacer size={8} />
            <H1>{locales.title}</H1>
            <VSpacer size={8} />
            <ScrollView>
              <Body>{locales.body}</Body>
              <Link
                onPress={presentWhatIsPspBottomSheet}
                testID={"whatIsPSPTestID"}
              >
                {locales.link}
              </Link>
              <VSpacer size={24} />
              <RadioListHeader
                leftColumnTitle={locales.leftColumnTitle}
                rightColumnTitle={locales.rightColumnTitle}
              />
              <VSpacer size={8} />
              <RadioButtonList<IOPayPalPsp["id"]>
                key="paypal_psp_selection"
                items={getPspListRadioItems(pspList)}
                selectedItem={selectedPsp?.id}
                onPress={(idPsp: string) => {
                  setSelectedPsp(pspList.find(p => p.id === idPsp));
                }}
              />
            </ScrollView>
          </View>
          <FooterWithButtons
            type={"TwoButtonsInlineThird"}
            leftButton={buttonsProps.cancelButtonProps}
            rightButton={{
              ...buttonsProps.continueButtonProps,
              disabled: selectedPsp === undefined
            }}
          />
          {bottomSheet}
        </SafeAreaView>
      ) : (
        <LoadingErrorComponent
          testID={"PayPalPpsSelectionScreenLoadingError"}
          isLoading={!isError(props.pspList)}
          loadingCaption={I18n.t("global.remoteStates.loading")}
          onRetry={searchPaypalPsp}
        />
      )}
    </BaseScreenComponent>
  );
};

const mapDispatchToProps = (dispatch: Dispatch) => ({
  goBack: () => dispatch(walletAddPaypalBack()),
  cancel: () => dispatch(walletAddPaypalCancel()),
  setPspSelected: (psp: IOPayPalPsp) =>
    dispatch(walletAddPaypalPspSelected(psp))
});
const mapStateToProps = (state: GlobalState) => ({
  pspList: payPalPspSelector(state)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(PayPalPspSelectionScreen);
