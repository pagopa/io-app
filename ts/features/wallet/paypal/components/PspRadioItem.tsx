// component that represents the item in the radio list
import React from "react";
import { Dimensions, Image, StyleSheet } from "react-native";
import { View } from "native-base";
import I18n from "../../../../i18n";
import { useIOBottomSheetRaw } from "../../../../utils/bottomSheet";
import { useImageResize } from "../../onboarding/bancomat/screens/hooks/useImageResize";
import { H4 } from "../../../../components/core/typography/H4";
import TouchableDefaultOpacity from "../../../../components/TouchableDefaultOpacity";
import IconFont from "../../../../components/ui/IconFont";
import { IOColors } from "../../../../components/core/variables/IOColors";
import { PayPalPsp } from "../screen/PayPalPspSelectionScreen";
import { PspInfoBottomSheetContent } from "./PspInfoBottomSheet";

const PSP_LOGO_MAX_WIDTH = Dimensions.get("window").width;
const PSP_LOGO_MAX_HEIGHT = 32;
type RadioItemProps = {
  psp: PayPalPsp;
};

const styles = StyleSheet.create({
  pspLogo: {
    height: 32,
    flex: 1,
    flexDirection: "row",
    justifyContent: "flex-start"
  },
  radioItemBody: {
    alignItems: "flex-start",
    justifyContent: "space-between",
    flexDirection: "row"
  },
  radioItemRightContainer: {
    flex: 1,
    flexDirection: "column",
    alignItems: "flex-end"
  },
  radioItemRight: {
    flex: 1,
    flexDirection: "column",
    alignItems: "flex-end"
  }
});

/**
 * item used in a radio list
 * it represent a PayPal psp with logo and info icon
 * @param props
 * @constructor
 */
export const PspRadioItem = (
  props: RadioItemProps
): React.ReactElement | null => {
  const { psp } = props;
  const imgDimensions = useImageResize(
    PSP_LOGO_MAX_WIDTH,
    PSP_LOGO_MAX_HEIGHT,
    psp.logoUrl
  );
  const pspInfoBottomSheet = useIOBottomSheetRaw(
    Math.min(470, Dimensions.get("window").height)
  );
  const handleInfoPress = () => {
    void pspInfoBottomSheet.present(
      <PspInfoBottomSheetContent
        onButtonPress={pspInfoBottomSheet.dismiss}
        pspFee={psp.fee}
        pspName={psp.name}
        pspPrivacyUrl={psp.privacyUrl}
      />,
      I18n.t("wallet.onboarding.paypal.selectPsp.infoBottomSheet.title", {
        pspName: psp.name
      })
    );
  };
  return (
    <View style={styles.radioItemBody}>
      {/* show the psp name while its image is loading */}
      {imgDimensions.fold<React.ReactNode>(
        <H4 weight={"SemiBold"} color={"bluegreyDark"}>
          {psp.name}
        </H4>,
        imgDim => (
          <Image
            source={{ uri: psp.logoUrl }}
            style={[styles.pspLogo, { width: imgDim[0], height: imgDim[1] }]}
            resizeMode={"contain"}
          />
        )
      )}
      <View style={styles.radioItemRightContainer}>
        <TouchableDefaultOpacity
          onPress={handleInfoPress}
          style={styles.radioItemRight}
        >
          <IconFont name={"io-info"} size={24} color={IOColors.blue} />
        </TouchableDefaultOpacity>
      </View>
    </View>
  );
};
