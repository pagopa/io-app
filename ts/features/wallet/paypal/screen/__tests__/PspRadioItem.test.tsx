import { NonNegativeNumber } from "@pagopa/ts-commons/lib/numbers";
import { fireEvent, render } from "@testing-library/react-native";
import React from "react";
import { PspRadioItem } from "../../components/PspRadioItem";
import { PayPalPsp } from "../PayPalPspSelectionScreen";
import { privacyUrl } from "../../../../../config";

const mockPresent = jest.fn();
jest.mock("@gorhom/bottom-sheet", () => ({
  useBottomSheetModal: () => ({
    present: mockPresent
  })
}));

const payPalPsp: PayPalPsp = {
  id: "1",
  logoUrl: "https://paytipper.com/wp-content/uploads/2021/02/logo.png",
  name: "PayTipper",
  fee: 100 as NonNegativeNumber,
  privacyUrl,
  tosUrl: privacyUrl
};

describe("PspRadioItem", () => {
  jest.useFakeTimers();
  it(`match snapshot`, () => {
    const component = render(<PspRadioItem psp={payPalPsp} />);
    expect(component).toMatchSnapshot();
  });

  it(`component should be defined`, () => {
    const component = render(
      <PspRadioItem psp={payPalPsp} testID={"PspRadioItemTestID"} />
    );
    expect(component.queryByTestId("PspRadioItemTestID")).not.toBeNull();
  });

  it(`should be present the info icon`, () => {
    const component = render(<PspRadioItem psp={payPalPsp} />);
    expect(component.queryByTestId("infoIconTestID")).not.toBeNull();
  });

  it(`should be present at least the name or the logo`, () => {
    const component = render(<PspRadioItem psp={payPalPsp} />);
    const pspName = component.queryByTestId("pspNameTestID");
    const pspLogo = component.queryByTestId("pspNameLogoID");
    expect(pspName === null && pspLogo === null).toBeFalsy();
  });

  it(`info icon should present the bottom sheet`, () => {
    const component = render(<PspRadioItem psp={payPalPsp} />);
    const infoIcon = component.queryByTestId("infoIconTestID");
    if (infoIcon) {
      fireEvent.press(infoIcon);
      expect(mockPresent).toHaveBeenCalledTimes(1);
    }
  });
});
