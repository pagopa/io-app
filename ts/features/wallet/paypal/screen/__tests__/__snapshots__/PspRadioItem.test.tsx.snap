// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PspRadioItem match snapshot 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
    ]
  }
>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "600",
        },
      ]
    }
    testID="pspNameTestID"
    weight="SemiBold"
    weightColorFactory={[Function]}
  >
    PayTipper
  </Text>
  <View
    style={
      Array [
        Object {},
        Object {
          "alignItems": "flex-end",
          "flex": 1,
          "flexDirection": "column",
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="infoIconTestID"
    >
      <Text
        accessibilityElementsHidden={true}
        accessibilityLabel=""
        accessible={false}
        allowFontScaling={false}
        importantForAccessibility="no-hide-descendants"
        style={
          Array [
            Object {
              "color": "#0073E6",
              "fontSize": 24,
            },
            Array [
              Object {},
              Object {
                "color": "#0073E6",
              },
              Object {
                "fontSize": 24,
              },
            ],
            Object {
              "fontFamily": "io-icon-font",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        î¤¦
      </Text>
    </View>
  </View>
</View>
`;
